x-pack/metricbeat/module/mssql/fetcher.go

```diff
@@ -0,0 +1,176 @@
+// it can be extracted to a common function
+func (f *Fetcher) Report(reporter mb.ReporterV2) {
+	// Run queries concurrently.
+	f.doQueries()
+
+	if f.Error != nil {
+		reporter.Error(f.Error)
+		return
+	}
+
+	for _, e := range f.Results {
+		reporter.Event(mb.Event{
+			MetricSetFields: e,
+		})
+	}
+}
+
+// doQueries is executed on object creation from the metricsets via NewFetcher.
+// It will launch a goroutine for each query stored and wait for all results
+// before returning.
+func (f *Fetcher) doQueries() {
+	f.wg.Add(len(f.queries))
+	for _, q := range f.queries {
+		go func() {
+			defer f.wg.Done()
+			f.doEventsFetcherQuery(q)
+		}()
+	}
+	f.wg.Wait()
+}
+
+// doEventsFetcherQuery adds common.Mapr objects concurrently.
+func (f *Fetcher) doEventsFetcherQuery(q string) {
+	results, err := f.getEventsWithQuery(f.db, q)
+	if err != nil {
+		f.resultsLock.Lock()
+		defer f.resultsLock.Unlock()
+
+		f.errs = append(f.errs, err)
+		return
+	}
+
+	f.resultsLock.Lock()
+	defer f.resultsLock.Unlock()
+	f.Results = append(f.Results, results...)
+}
+
+// getEventsWithQuery performs the query on the database and converts the rows
+// to an slice of common.MapStr.
+func (f *Fetcher) getEventsWithQuery(db *sql.DB, query string) ([]common.MapStr, error) {
+	// Returns the global status, also for versions previous 5.0.2
+	rows, err := db.Query(query)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error performing db query='%v'", query)
+	}
+	defer rows.Close()
+
+	results, err := f.eventsMapping(rows)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not convert row result to beats mapstr")
+	}
+
+	return results, nil
+}
+
+// eventsMapping takes a *sql.Rows to convert them to a []common.MapStr slice
+// dynamically without knowing the column names in advance.
+func (f *Fetcher) eventsMapping(rows *sql.Rows) ([]common.MapStr, error) {
+	columnNames, err := rows.Columns()
+	if err != nil {
+		return nil, errors.Wrap(err, "error getting column names")
+	}
+
+	var results []common.MapStr
+	dest := make([]interface{}, len(columnNames))
+
+	for rows.Next() {
+		// We assign pointers to the destination to pass it to Scan.
+		rawResult := make([]*string, len(columnNames))
+		for i := range rawResult {
+			dest[i] = &rawResult[i]
+		}
+
+		if err = rows.Scan(dest...); err != nil {
+			return nil, errors.Wrap(err, "error scanning row of result")
+		}
+
+		// Now we need to get the values of the pointers back into a normal
+		// map[string]interface{} to use it with the schema.
+		mapOfResults := make(map[string]interface{})
+
+		for i, res := range rawResult {
+			if res != nil {
+				mapOfResults[columnNames[i]] = *res
+			}
+		}
+
+		result, err := f.schema.Apply(mapOfResults)
+		if err != nil {
+			// TODO: Handle schema error.
+			f.log.Debugw("error applying schema to query result", "error", err)
+		}
+
+		f.resultsLock.Lock()
+		results = append(results, result)
+		f.resultsLock.Unlock()
```

Comment:
Is there a need for mutex here?

Reply:
Oh, I see it is used when appending to `f.Results`, not sure in any case if a mutex is needed here as the append is already in a mutex.

Another option for this case could be to have a results channel, then there wouldn't be a need for any lock.

Or directly report the events. Is there any reason to keep them in a results list?

Reply:
I'll consider the option you describe slowly. The problem I have with channels in this context is that I prefer to use them when their lifetime can be controlled end to end within the source code (no frameworks involved) and / or it's a complex use case, not like this which is rather simple. I have seen that it's easier to people to reason about mutexes.

Reply:
Well, in this case I think that it could be simpler with a channel, and "share memory by communicating" is a golang principle :)

In any case we can directly report the events once they are generated if we pass the reporter till `eventMapping` as we do in other modules, and then there is no need for mutexes nor channels.